qiling/debugger/qdb/utils.py

```diff
@@ -36,81 +36,69 @@
 
 
-def qdb_print(msgtype: QDB_MSG, msg: str) -> None:
-    """
-    color printing
-    """
+_K = TypeVar('_K')
+_V = TypeVar('_V')
 
-    def print_error(msg):
-        return f"{color.RED}[!] {msg}{color.END}"
 
-    def print_info(msg):
-        return f"{color.CYAN}[+] {msg}{color.END}"
+def qdb_print(level: QDB_MSG, msg: str) -> None:
+    """Log printing.
+    """
 
-    color_coated = {
-            QDB_MSG.ERROR: print_error,
-            QDB_MSG.INFO : print_info,
-            }.get(msgtype)(msg)
+    decorations = {
+        QDB_MSG.ERROR: ('!', color.RED),
+        QDB_MSG.INFO : ('+', color.CYAN),
+    }
 
-    print(color_coated)
+    tag, col = decorations[level]
 
+    print(f'{col}[{tag}] {msg}{color.END}')
 
-def setup_address_marker():
 
-    class Marker:
-        """provide the ability to mark an address as a more easier rememberable alias
-        """
+class Marker:
+    """provide the ability to mark an address as a more easier rememberable alias
+    """
 
-        def __init__(self):
-            self._mark_list = {}
+    def __init__(self):
+        self._mark_list: Dict[str, int] = {}
 
-        def get_symbol(self, sym):
-            """
-            get the mapped address to a symbol if it's in the mark_list
-            """
+    def get_address(self, sym: str) -> Optional[int]:
+        """
+        get the mapped address to a symbol if it's in the mark_list
+        """
 
-            return self._mark_list.get(sym, None)
+        return self._mark_list.get(sym)
 
-        @property
-        def mark_list(self):
-            """
-            get a list about what we marked
-            """
+    @property
+    def mark_list(self):
+        """
+        get a list about what we marked
+        """
 
-            return self._mark_list.items()
+        return self._mark_list.items()
 
-        def gen_sym_name(self):
-            """
-            generating symbol name automatically
-            """
+    def gen_sym_name(self) -> str:
+        """
+        generating symbol name automatically
+        """
 
-            sym_name, idx = "sym0", 0
-            while sym_name in self._mark_list:
-                idx += 1
-                sym_name = f"sym{idx}"
+        syms = len(self._mark_list)
 
-            return sym_name
+        # find the next available 'sym#'
+        return next((f'sym{i}' for i in range(syms) if f'sym{i}' not in self._mark_list), f'sym{syms}')
 
-        def mark_only_loc(self, loc):
-            """
-            mark when location provided only
-            """
+    def mark(self, loc: int, sym: Optional[str] = None) -> str:
+        """
+        mark loc as sym
+        """
 
-            sym_name = self.gen_sym_name()
-            self.mark(sym_name, loc)
-            return sym_name
+        sym = sym or self.gen_sym_name()
 
-        def mark(self, sym: str, loc: int):
-            """
-            mark loc as sym
-            """
+        if sym in self.mark_list:
```

Comment:
Also, this will make duplicated symbol detection never happen.

Reply:
I am not sure what kind of change is required here (if any)

Reply:
The function do_mark will detect if the user-provided symbol name duplicated or not. But, the `self.mark_list` is actually an items object, therefore `sym in self.mark_list` will always be false.